apps/examples/rtc/rtc_main.c

```diff
@@ -150,10 +161,230 @@ static void *rtc_test(void *arg)
 	return NULL;
 }
 
+#ifdef CONFIG_RTC_ALARM
+/****************************************************************************
+ * Name: alarm_handler
+ ****************************************************************************/
+
+static void alarm_handler(int signo, FAR siginfo_t *info, FAR void *ucontext)
+{
+	int ret;
+	int fd;
+	int secs;
+	int almndx;
+	FAR struct rtc_time alarm_recieve_time;
+
+	fd = open(RTC_DEVPATH, O_WRONLY);
+	if (fd < 0) {
+		fprintf(stderr, "ERROR: Failed to open %s: %d\n", RTC_DEVPATH, errno);
+		return;
+	}
+	almndx = info->si_value.sival_int;
+	if (almndx >= 0 && almndx < CONFIG_RTC_NALARMS) {
+		ret = ioctl(fd, RTC_RD_TIME, (unsigned long)&alarm_recieve_time);
+		if (ret < 0) {
+			fprintf(stderr, "ERROR: RTC_RD_TIME ioctl failed: %d\n", errno);
+			close(fd);
+			return;
+		}
+		if(g_relset) {
+			secs = (alarm_recieve_time.tm_hour - g_alarm_time.tm_hour) * 3600 + (alarm_recieve_time.tm_min - g_alarm_time.tm_min) * 60 + (alarm_recieve_time.tm_sec - g_alarm_time.tm_sec);
+			printf("Alarm received after %d seconds\n", secs);
+		} else {
+			printf("Alarm recieved at %d:%d:%d\n",alarm_recieve_time.tm_hour, alarm_recieve_time.tm_min, alarm_recieve_time.tm_sec);
+		}
+	}
+	close(fd);
+}
+
+/****************************************************************************
+ * Name: show_usage
+ ****************************************************************************/
+
+static void show_usage(FAR const char *progname)
+{
+	fprintf(stderr, "\nUsage:  rtc\n");
+	fprintf(stderr, "\tor: rtc stop\n");
+	fprintf(stderr, "\t\tStart, or Stop rtc test\n");
+	fprintf(stderr, "\tor rtc alarm [-r seconds] [-a hh:mm:ss] [-c]\n");
+	fprintf(stderr, "\t\tset, or cancel rtc alarm\n");
+	fprintf(stderr, "Where:\n");
+	fprintf(stderr, "\t-a hh:mm:ss \n");
+	fprintf(stderr, "\t\tset alarm by absolute time in hh:mm:ss format.\n");
+	fprintf(stderr, "\t-r <seconds> \n");
+	fprintf(stderr, "\t\tset alarm by relative time. The number of seconds until the alarm expires.\n");
+	fprintf(stderr, "\t-c\n\t\tCancel previously set alarm\n");
+	fprintf(stderr, "\t\t");
+}
+
+int rtc_alarm_main(int argc, char *argv[])
+{
+	int seconds = 0;
+	int mins = 0;
+	int hrs = 0;
+	bool badarg = false;
+	bool cancelmode = false;
+	int fd;
+	int ret;
+	int opt;
+	int cur_time;
+	int user_time;
+	FAR struct sigaction act;
+	FAR struct rtc_setalarm_s setalarm;
+	FAR	struct rtc_setrelative_s setrel;
+	sigset_t set;
+
+	while((opt = getopt(argc, argv, ":a:r:c")) != ERROR) {
+		switch (opt) {
+			case 'a':
+				/* -a: alarm is set by absolute time (RTC_SET_ALARM ioctl call).
+				 */
+				g_relset = false;
+				break;
+			case 'r':
+				/* -r: alarm is set by relative time (RTC_SET_RELATIVE ioctl call).
+				 */
+				g_relset = true;
+				break;
+			case 'c':
+				cancelmode = true;
+				break;
+			default:
+				fprintf(stderr, "<unknown parameter '-%c'>\n\n", opt);
+				/* fall through */
+
+			case '?':
+			case ':':
+				badarg = true;
```

Comment:
Do you need three different global variables?
```
switch (opt) {
case 'a':
   command_mode = ABS_TIME_MODE;
   break;
case 'r':
   command_mode = REL_TIME_MODE;
   break;
case '?':
case ':':
   command_mode = INVALID;
   break;
}

switch (command_mode) {
case INVALID:
   printf("invalid command");
   // call help message;
   ret = ERROR;
   break;
case ABS_....:
   // call function for abs
   ret = OK;
   break;
case ...
```
How about this? This function looks too long.

Reply:
Thank you for the check and suggestion.. I also think it will be much better and I will update it.